grails-app/assets/javascripts/facet-search.js

```diff
@@ -0,0 +1,223 @@
+            }
+        },
+
+        onSelect: function(suggestion) {
+            facetsMap[this.id] = suggestion;
+        }
+    });
+});
+
+function facetSearch() {
+    var baseURL = BC_CONF.contextPath + '/occurrences/search';
+    var commonArguments = 'sort=first_loaded_date&dir=desc#map';
+    var inputs = [].slice.call(document.getElementsByClassName('js-search-input'));
+
+    var textInputs = inputs.filter(function(input) {
+        return input.dataset.queryParam === 'q';
+    });
+
+    // Facet inputs that are not taxon auto-complete inputs.
+    var facetInputs = inputs.filter(function(input) {
+        return input.dataset.queryParam === 'fq' && input.dataset.facetName !== 'taxon_name';
+    });
+
+    // Facet inputs that are taxon auto-complete inputs.
+    var taxonFacetInputs = inputs.filter(function(input) {
+        return input.dataset.queryParam === 'fq' && input.dataset.facetName === 'taxon_name';
+    });
+
+    // We currently use just one q input and query.
+    var textQuery = 'q=' + (textInputs[0].value || '*:*');
+    var facetArgs = facetInputs.map(function(input) {
+        var facetName = input.dataset.facetName;
+        var facet = facetsMap[input.id];
+        var value = input.value;
+
+        if(facetName === 'institution_collection') {
+            return facetCollectionParam(value);
+        } else if(facetName === 'start_date') {
+            return facetDateParam();
+        } else if(facetName === 'end_date') {
+            // Do nothing: date handler takes care of them both when dealing with the start date.
+        } else {
+            return value ? encodeURIComponent(facetName + ':"' + value + '"') : null;
+        }
+    }).filter(function(arg) {
+        return arg !== null;
+    });
+
+    var taxonFacetArgs = taxonFacetInputs.map(function(input) {
+        var facet = facetsMap[input.id];
+
+        return facet ? facetTaxonParam(facet) : null;
+    }).filter(function(arg) {
+        return arg !== null;
+    });
+
+    // Really? Yes, this is how they eat.
+    var facetsQuery = facetArgs.length ? 'fq=' + facetArgs.join('&fq=') : '';
+    var taxonFacetQuery = taxonFacetArgs.length ? 'fq=(' + taxonFacetArgs.join(' OR ') + ')' : '';
+    var searchURL = [baseURL + '?' + textQuery, taxonFacetQuery, facetsQuery, commonArguments].filter(function(part) {
+        return part !== '';
+    }).join('&');
+
+    window.location.assign(searchURL);
+}
+
+function clearSearchInputs() {
+    var inputs = [].slice.call(document.getElementsByClassName('js-search-input'));
+
+    // Clear input values.
+    inputs.forEach(function(input) {
+        input.value = '';
+    });
+
+    // Clear auto-complete values.
+    Object.keys(facetsMap).forEach(function(key) {
+        facetsMap[key] = null;
+    });
+}
+
+function facetTaxonParam(facet) {
+    var rank = facet.data.rank;
+    var taxon = facet.data[rank] || facet.data.scientificName; // not sure the second half of this is necessary
```

Comment:
Oh it is. When doing taimeatlas app, had to learn it the hard way - some taxons are missing rank entries for their own rank.